File: content/develop/use-cases/feature-store/rust/build_features.rs
//! Synthesize a small batch of users with realistic-looking features
//! and bulk-load them into Redis with a 24-hour key-level TTL.
//!
//! Stands in for the nightly Spark / Feast materialization job in a
//! real deployment. In production the equivalent of this script lives
//! in an offline pipeline that reads from the offline store and
//! writes the serving-time hashes into Redis via `HSET` + `EXPIRE`.

use std::collections::BTreeMap;

use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};

use crate::feature_store::{FeatureMap, FeatureValue};

const COUNTRY_CHOICES: &[&str] = &[
    "US", "GB", "DE", "FR", "IN", "BR", "JP", "AU", "CA", "NL",
];
const RISK_SEGMENTS: &[&str] = &["low", "medium", "high"];
const RISK_WEIGHTS: &[u32] = &[70, 25, 5];
const CHARGEBACK_BUCKETS: &[i64] = &[0, 1, 2, 3];
const CHARGEBACK_WEIGHTS: &[u32] = &[85, 10, 4, 1];

/// Generate `count` synthetic user feature rows. The shape mirrors a
/// small fraud-scoring feature set: country and risk segment as
/// TAG-like categorical features, plus a few numeric aggregates over
/// recent windows.
pub fn synthesize_users(count: usize, seed: u64) -> Vec<(String, FeatureMap)> {
    let mut rng = StdRng::seed_from_u64(seed);
    let mut users = Vec::with_capacity(count);
    for i in 1..=count {
        let uid = format!("u{:04}", i);
        let mut row: FeatureMap = BTreeMap::new();
        row.insert("country_iso".into(), FeatureValue::Str(
            COUNTRY_CHOICES[rng.gen_range(0..COUNTRY_CHOICES.len())].into()));
        row.insert("risk_segment".into(), FeatureValue::Str(
            weighted_str(&mut rng, RISK_SEGMENTS, RISK_WEIGHTS).into()));
        row.insert("account_age_days".into(), FeatureValue::Int(rng.gen_range(7..=2400)));
        row.insert("tx_count_7d".into(), FeatureValue::Int(rng.gen_range(0..=80)));
        let avg = (rng.gen_range(5.0..350.0_f64) * 100.0).round() / 100.0;
        row.insert("avg_amount_30d".into(), FeatureValue::Float(avg));
        row.insert(
            "chargeback_count_180d".into(),
            FeatureValue::Int(weighted_int(&mut rng, CHARGEBACK_BUCKETS, CHARGEBACK_WEIGHTS)),
        );
        users.push((uid, row));
    }
    users
}

fn weighted_str<R: Rng + ?Sized>(rng: &mut R, items: &[&'static str], weights: &[u32]) -> &'static str {
    let total: u32 = weights.iter().sum();
    let mut r = rng.gen_range(0..total);
    for (i, w) in weights.iter().enumerate() {
        if r < *w { return items[i]; }
        r -= w;
    }
    items[items.len() - 1]
}

fn weighted_int<R: Rng + ?Sized>(rng: &mut R, items: &[i64], weights: &[u32]) -> i64 {
    let total: u32 = weights.iter().sum();
    let mut r = rng.gen_range(0..total);
    for (i, w) in weights.iter().enumerate() {
        if r < *w { return items[i]; }
        r -= w;
    }
    items[items.len() - 1]
}

/// CLI entry point invoked by `build_features_bin.rs`. Parses flags,
/// opens a one-shot `ConnectionManager`, and calls `bulk_load`.
pub async fn cli_main() -> redis::RedisResult<()> {
    let mut redis_url = "redis://127.0.0.1/".to_string();
    let mut count: usize = 200;
    let mut ttl_seconds: u64 = 24 * 60 * 60;
    let mut key_prefix = "fs:user:".to_string();
    let mut seed: u64 = 42;

    let args: Vec<String> = std::env::args().skip(1).collect();
    let mut i = 0usize;
    while i < args.len() {
        match args[i].as_str() {
            "--redis-url" => { redis_url = args[i + 1].clone(); i += 2; }
            "--count" => { count = args[i + 1].parse().unwrap_or(count); i += 2; }
            "--ttl-seconds" => { ttl_seconds = args[i + 1].parse().unwrap_or(ttl_seconds); i += 2; }
            "--key-prefix" => { key_prefix = args[i + 1].clone(); i += 2; }
            "--seed" => { seed = args[i + 1].parse().unwrap_or(seed); i += 2; }
            "-h" | "--help" => {
                println!("Usage: build_features [--redis-url URL] [--count N] [--ttl-seconds S] [--key-prefix PREFIX] [--seed N]");
                return Ok(());
            }
            other => {
                eprintln!("Unknown argument: {other}");
                std::process::exit(2);
            }
        }
    }

    let client = redis::Client::open(redis_url.as_str())?;
    let conn = redis::aio::ConnectionManager::new(client).await?;
    let store = crate::feature_store::FeatureStore::new(
        conn,
        &key_prefix,
        ttl_seconds,
        crate::feature_store::DEFAULT_STREAMING_TTL_SECONDS,
    );

    let rows = synthesize_users(count, seed);
    let loaded = store.bulk_load(&rows, ttl_seconds).await?;
    println!(
        "Materialized {} users at {}* with a {}s key-level TTL.",
        loaded, key_prefix, ttl_seconds
    );
    Ok(())
}

File: content/develop/use-cases/feature-store/dotnet/BuildFeatures.cs
using StackExchange.Redis;

namespace FeatureStoreDemo;

/// <summary>
/// Synthesize a small batch of users with realistic-looking features
/// and bulk-load them into Redis with a 24-hour key-level TTL.
/// </summary>
/// <remarks>
/// Stands in for the nightly Spark / Feast materialization job in a
/// real deployment. In production the equivalent of this script lives
/// in an offline pipeline that reads from the offline store and
/// writes the serving-time hashes into Redis via <c>HSET</c> +
/// <c>EXPIRE</c>.
/// </remarks>
public static class BuildFeatures
{
    private static readonly string[] CountryChoices = {
        "US", "GB", "DE", "FR", "IN", "BR", "JP", "AU", "CA", "NL",
    };
    private static readonly string[] RiskSegments = { "low", "medium", "high" };
    private static readonly int[] RiskWeights = { 70, 25, 5 };
    private static readonly int[] ChargebackBuckets = { 0, 1, 2, 3 };
    private static readonly int[] ChargebackWeights = { 85, 10, 4, 1 };

    /// <summary>
    /// Generate <paramref name="count"/> synthetic user feature rows.
    /// </summary>
    public static Dictionary<string, IReadOnlyDictionary<string, object>> SynthesizeUsers(
        int count, int seed)
    {
        var rng = new Random(seed);
        var users = new Dictionary<string, IReadOnlyDictionary<string, object>>(count);
        for (int i = 1; i <= count; i++)
        {
            var uid = $"u{i:D4}";
            users[uid] = new Dictionary<string, object>
            {
                ["country_iso"] = CountryChoices[rng.Next(CountryChoices.Length)],
                ["risk_segment"] = WeightedChoice(rng, RiskSegments, RiskWeights),
                ["account_age_days"] = 7 + rng.Next(2394),
                ["tx_count_7d"] = rng.Next(81),
                ["avg_amount_30d"] = Math.Round(5.0 + rng.NextDouble() * 345.0, 2),
                ["chargeback_count_180d"] = WeightedChoiceInt(rng, ChargebackBuckets, ChargebackWeights),
            };
        }
        return users;
    }

    /// <summary>
    /// CLI entry point. Run with:
    /// <c>dotnet run --project . -- --mode build-features --count 500</c>
    /// </summary>
    public static async Task<int> RunCliAsync(string[] args)
    {
        var redisUri = "localhost:6379";
        var count = 200;
        var ttlSeconds = 24L * 60L * 60L;
        var keyPrefix = "fs:user:";
        var seed = 42;

        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "--redis-uri" when i + 1 < args.Length:
                    redisUri = args[++i]; break;
                case "--count" when i + 1 < args.Length:
                    count = int.Parse(args[++i]); break;
                case "--ttl-seconds" when i + 1 < args.Length:
                    ttlSeconds = long.Parse(args[++i]); break;
                case "--key-prefix" when i + 1 < args.Length:
                    keyPrefix = args[++i]; break;
                case "--seed" when i + 1 < args.Length:
                    seed = int.Parse(args[++i]); break;
                case "-h":
                case "--help":
                    Console.WriteLine(
                        "Usage: dotnet run -- --mode build-features [--redis-uri URI] " +
                        "[--count N] [--ttl-seconds S] [--key-prefix PREFIX] [--seed N]");
                    return 0;
            }
        }

        var mux = await ConnectionMultiplexer.ConnectAsync(redisUri);
        try
        {
            var store = new FeatureStore(mux, keyPrefix, ttlSeconds,
                FeatureStore.DefaultStreamingTtlSeconds);
            var rows = SynthesizeUsers(count, seed);
            var loaded = await store.BulkLoadAsync(rows, ttlSeconds);
            Console.WriteLine(
                $"Materialized {loaded} users at {keyPrefix}* with a {ttlSeconds}s key-level TTL.");
        }
        finally
        {
            await mux.CloseAsync();
        }
        return 0;
    }

    private static string WeightedChoice(Random rng, string[] items, int[] weights)
    {
        int total = 0;
        foreach (var w in weights) total += w;
        int r = rng.Next(total);
        for (int i = 0; i < items.Length; i++)
        {
            r -= weights[i];
            if (r < 0) return items[i];
        }
        return items[^1];
    }

    private static int WeightedChoiceInt(Random rng, int[] items, int[] weights)
    {
        int total = 0;
        foreach (var w in weights) total += w;
        int r = rng.Next(total);
        for (int i = 0; i < items.Length; i++)
        {
            r -= weights[i];
            if (r < 0) return items[i];
        }
        return items[^1];
    }
}
